File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def check(self, mid, piles, h):
        spend = 0 
        for i in range(len(piles)):
            q = piles[i] // mid
            r = piles[i] % mid
            
            if r > 0:
                q += 1
            spend += q
            
        return spend <= h
            
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        
        left = 0
        right = 10**9
        
        while left+1 < right:
            mid = left + ((right - left) // 2)
            
            if self.check(mid, piles, h):
                right = mid
            else:
                left = mid
        
        return right
        

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int left = 0;
        int right = (int) Math.pow(10, 9);

        while(left+1 < right){
            int mid = left + ((right - left) / 2);

            if(check(mid,piles,h)) {
                right = mid;
            } else {
                left = mid;
            }
        }
        return right;
    }
     private boolean check(int mid, int[] piles, int h) {
        long spendHour = 0;
        for (int pile : piles) {
            int q = Math.floorDiv(pile, mid);
            int r = Math.floorMod(pile, mid);

            if(r > 0) {
                q++;
            }
            spendHour += q;
        }
       return spendHour <= (long) h;
    }
}